### DISCUSSION TITLE
How to use remark-footnotes with react-markdown?

### DISCUSSION THREAD

**Discussion by damiensnyder** [remarkjs/remark#554] [Q&A] [answered]
I am using react-markdown to integrate markdown into webpages served by Next.js. Already I have found, from the examples on the readme, how to use the remark-math plugin with react-katex. However, there is no such example for remark-footnotes that I can find anywhere. My intuition is that the syntax to import and use the footnote plugin should be similar, but I'm not sure what specifically to import and what to include in `renderers` and `plugins`. How do I use these two libraries together?

**Comment by wooorm** [comment, marked as answer]
> My intuition is that the syntax to import and use the footnote plugin should be similar

It is indeed exactly the same, but now using remark-footnotes instead of remark-math.

> but I'm not sure what specifically to import and what to include in renderers and plugins. How do I use these two libraries together?

You can try it out, and you’ll get an error because some custom nodes aren’t handled.

```js
import React from 'react'
import ReactMarkdown from 'react-markdown'
import {render} from 'react-dom'
import footnotes from 'remark-footnotes'

const renderers = {
}

render(
  <ReactMarkdown
    plugins={[footnotes]}
    renderers={renderers}
    children={`place some footnotes here`}
  />,
  document.body
)
// This throws, because the footnotes can’t be handled.
```

From there, you can add a renderer:

```js
const renderers = {
  typeOfNodeThatFailed: function () {
    console.log(arguments)
    return ''
  }
}
```

And go from there.

**Reply by damiensnyder** [reply]
Thank you for your response! I added `footnotes` as a plugin, which I'd probably tried before, but I hadn't added a renderer because I didn't know what to add. Now I've realized that I can pass custom footnotes into `renderers.footnoteReference` and `renderers.footnoteDefinition`, and it works now.

**Reply by wooorm** [reply]
Sweet, glad to hear!

**Reply by wooorm** [reply]
To add: we can’t add *every* plugin to the docs, as that would be a hundred examples. (See also: https://github.com/remarkjs/react-markdown/issues/479). But if you have a suggestion on how to make your exact case easier through docs without explicitly mentioning a specific plugin, let me know!

**Reply by szhshp** [reply]
> Thank you for your response! I added `footnotes` as a plugin, which I'd probably tried before, but I hadn't added a renderer because I didn't know what to add. Now I've realized that I can pass custom footnotes into `renderers.footnoteReference` and `renderers.footnoteDefinition`, and it works now.

Can you please share your solution? I still cannot figure out how to do that. Thanks

**Reply by damiensnyder** [reply]
[This page](https://github.com/damiensnyder/personal-website/blob/main/pages/blog/%5Bcode%5D.tsx) is where I am using it in production. The relevant code snippets are:

```
import footnotes from 'remark-footnotes';

export const markdownRenderers: any = {
  footnoteReference: FootnoteReference,
  footnoteDefinition: FootnoteDefinition
}
export const markdownPlugins: any[] = [footnotes];

function FootnoteReference(props): ReactElement {
  return (
    <sup id={"ref-" + props.identifier}>
      <a href={"#def-" + props.identifier}>{props.label}</a>
    </sup>
  );
}

function FootnoteDefinition(props): ReactElement {
  return (
    <div className={styles.footnoteDefinition} id={"def-" + props.identifier}>
      <a className={styles.backToRef}
              href={"#ref-" + props.identifier}>{props.label}</a>
      <div className={styles.footnoteBody}>{props.children}</div>
    </div>
  );
}
```

Then the `ReactMarkdown` JSX item is:

```
<ReactMarkdown renderers={markdownRenderers}
    plugins={markdownPlugins}
    source={props.text} />
```

All you need to know is the name of the elements you're rendering, and you can add them as keys in `markdownRenderers`. For the `remark-footnotes` plugin, that's `footnoteReference` and `footnoteDefinition`. For the KaTeX plugin, that's `math` and `inline-math`. For vanilla markdown, it's `heading`, `image`, etc. (I didn't know you could set custom renderers for ordinary markdown elements for a long time...)

**Reply by riencoertjens** [reply]
for anyone reading this, in the above example 
```
<ReactMarkdown renderers={markdownRenderers}
    plugins={markdownPlugins}
    source={props.text} />
```

the `renderers` prop should be renamed to `components` since [this change](https://github.com/remarkjs/react-markdown/blob/main/changelog.md#change-renderers-to-components)

**Reply by riencoertjens** [reply]
actually it seems `footnoteReference` and `footnoteDefinition` don't work anymore :/

**Reply by nick-inkeep** [reply]
This doesn't to work for me. Any ideas?

**Reply by nick-inkeep** [reply]
@riencoertjens  or @damiensnyder , were you able to get this working with the new react-markdown that uses 'components'?

**Reply by riencoertjens** [reply]
I think I did, but this is a 1,5 yo solution from 2 versions ago, if you look at the [changelog](https://github.com/remarkjs/react-markdown/blob/main/changelog.md#800---2022-01-17) it says `plugins` prop is deprecated for `remarkPlugins`, maybe that's your problem?
